pytorch_lightning/trainer/trainer.py

```diff
@@ -668,12 +668,9 @@ def _call_and_handle_interrupt(self, trainer_fn: Callable, *args: Any, **kwargs:
             **kwargs: keyword arguments to be passed to `trainer_fn`
         """
         try:
-            if isinstance(self.strategy, DDPSpawnStrategy):
-                spawn_output: _SpawnOutput = self.strategy.spawn(trainer_fn, *args, **kwargs)
-                self.strategy._recover_results_in_main_process(spawn_output, self)
-                return spawn_output.trainer_results
-            else:
-                return trainer_fn(*args, **kwargs)
+            kwargs["trainer"] = self
+            kwargs["strategy"] = self.strategy
+            return self.strategy.launch(trainer_fn, *args, **kwargs)
```

Comment:
this should be opt-in

why couldn't it be designed as something like
```py
if self.strategy.launcher:
    return self.strategy.launch(trainer_fn, *args, **kwargs)
else:
    trainer_fn(*args, **kwargs)
```

then we have no need for the singleprocess launcher. single-device strategies and parallel plugins don't need it at all. 

wouldn't that make it easier to merge the DDP and DDP spawn strategy classes over time?

Reply:
another concern is on the cyclic dependency between a launcher and a strategy. if the launcher is part of the strategy, but the launcher also requires the strategy, doesn't this point to something we can clear up?

Reply:
since the launcher is not a class member variable, we can't really tell still this point whether it will execute on a single process or on a multi-device for DDP. Do you think it should be an argument? Then SingleProcessLauncher can be cleaned up.